src/pip/_internal/req/req_install.py

```diff
@@ -420,12 +421,18 @@ def check_if_exists(self, use_user_site):
         """
         if self.req is None:
             return
+
         # get_distribution() will resolve the entire list of requirements
         # anyway, and we've already determined that we need the requirement
         # in question, so strip the marker so that we don't try to
         # evaluate it.
         no_marker = Requirement(str(self.req))
         no_marker.marker = None
+
```

Comment:
Maybe we can also use the new `get_distribution()` function to rewrite the block below.

Reply:
I have trying to refactor the last remaining `pkg_resources.get_distribution()` at

https://github.com/pypa/pip/blob/f34f8d501ff03d8484ed0db7d4055cc9efcea6dd/src/pip/_internal/req/req_install.py#L429-L431

with the `get_distribution` function I have currently, but have been hitting a roadblock on getting a few functional tests to pass

I currently have the following:

```python
no_marker = Requirement(str(self.req))
no_marker.marker = None

# pkg_resources uses the canonical name to look up packages, but
# the name passed passed to get_distribution is not canonicalized
# so we have to explicitly convert it to a canonical name
no_marker.name = canonicalize_name(no_marker.name)
self.satisfied_by = get_distribution(str(no_marker))
if not self.satisfied_by:
    return
if self.req.specifier and self.satisfied_by.parsed_version \
        in self.req.specifier:
    if self.editable and self.satisfied_by:
        self.should_reinstall = True
        # when installing editables, nothing pre-existing should ever
        # satisfy
        self.satisfied_by = None
else:
    existing_dist = get_distribution(
        self.req.name
    )
    if use_user_site:
        if dist_in_usersite(existing_dist):
            self.should_reinstall = True
        elif (running_under_virtualenv() and
                dist_in_site_packages(existing_dist)):
            raise InstallationError(
                "Will not install to the user site because it will "
                "lack sys.path precedence to {} in {}".format(
                    existing_dist.project_name, existing_dist.location)
            )
    else:
        self.should_reinstall = True
```
Most of the functional tests are failing since `str(no_marker)`sometimes has the version specifier strings in them e.g. `INITools==0.3`, `argparse>=1.4` etc which calls the `get_installed_distributions` within `get_distribution` to fail  since that expects just the package name, after it it falls into the try/catch where `VersionConflict` is raised (which I am unsure how to catch and handle there)

Reply:
In addition to that I tried to replace 

```
no_marker = Requirement(str(self.req))
no_marker.marker = None
no_marker.name = canonicalize_name(no_marker.name)
self.satisfied_by = get_distribution(str(no_marker))
```

in above comment with 
```
self.satisfied_by = get_distribution(self.req.name)
```

but then saw that some of the tests are falsely passing. For e.g [tests.functional.test_install::test_upgrade_argparse_shadowed](https://github.com/pypa/pip/blob/master/tests/functional/test_install.py#L667)

the original `result.stdout` at the end of the test with `pkg_resources.get_distribution` is 

```
Collecting argparse>=1.4
  Using cached argparse-1.4.0-py2.py3-none-any.whl (23 kB)
Installing collected packages: argparse
  Attempting uninstall: argparse
    Found existing installation: argparse 1.3.0
    Uninstalling argparse-1.3.0:
      Successfully uninstalled argparse-1.3.0
Successfully installed argparse-1.4.0
```

but after the change, it shows up as `Requirement already satisfied: argparse>=1.4 in /Users/devesh/pip/.env/lib/python3.8/site-packages (1.3.0)`

I think there is still something missing in either my implementation of `get_distribution`, or the replacement of the `pkg_resources` exceptions with if/else, so any suggestion to fix those would be welcome.

Reply:
It seems like the typing is more complicated than I was imagining. It’s probably better to do this in a follow-up PR, this one is already getting too big.

Reply:
> It seems like the typing is more complicated than I was imagining.

Sorry, but could you let me know what do you mean when you say "typing is more complicated"?

>  It’s probably better to do this in a follow-up PR, this one is already getting too big.

Agreed, I will handle the last remaining `pkg_resources.get_distribution` in a follow-up PR. Once I do that, I would be really helpful if you could provide  suggestions from your end on the above questions on how to go about it.